plugins/SpectrumAnalyzer/SaProcessor.cpp

```diff
@@ -0,0 +1,556 @@
+								accL = (band_end - (int)band_end) * m_normSpectrumL[i];
+								accR = (band_end - (int)band_end) * m_normSpectrumR[i];
+							}
+						}
+					}
+					else
+					{
+						// Linear: always draws one or more pixels per band
+						for (target = band_start; target < band_end; target++)
+						{
+							if (target >= 0 && target < binCount())
+							{
+								pixel[target] = makePixel(m_normSpectrumL[i], m_normSpectrumR[i]);
+							}
+						}
+					}
+				}
+			}
+			#ifdef SA_DEBUG
+				// report FFT processing speed
+				start_time = std::chrono::high_resolution_clock::now().time_since_epoch().count() - start_time;
+				std::cout << "Processed " << m_framesFilledUp << " samples in " << start_time / 1000000.0 << " ms" << std::endl;
+			#endif
+
+			// clean up before checking for more data from input buffer
+			m_framesFilledUp = 0;
+		}
+	}
+}
+
+
+// Produce a spectrogram pixel from normalized spectrum data.
+// Values over 1.0 will cause the color components to overflow: this is left
+// intentionally untreated as it clearly indicates which frequency is clipping.
+// Gamma correction is applied to make small values more visible and to make
+// a linear gradient actually appear roughly linear. The correction should be
+// around 0.42 to 0.45 for sRGB displays (or lower for bigger visibility boost).
+QRgb SaProcessor::makePixel(float left, float right, float gamma_correction)
+{
+	if (m_controls->m_stereoModel.value())
+	{
+		float ampL = pow(left, gamma_correction);
+		float ampR = pow(right, gamma_correction);
+		return qRgb(m_controls->m_colorL.red() * ampL + m_controls->m_colorR.red() * ampR,
+					m_controls->m_colorL.green() * ampL + m_controls->m_colorR.green() * ampR,
+					m_controls->m_colorL.blue() * ampL + m_controls->m_colorR.blue() * ampR);
+	}
+	else
+	{
+		float ampL = pow(left, gamma_correction);
+		// make mono color brighter to compensate for the fact it is not summed
+		return qRgb(m_controls->m_colorMono.lighter().red() * ampL,
+					m_controls->m_colorMono.lighter().green() * ampL,
+					m_controls->m_colorMono.lighter().blue() * ampL);
+	}
+}
+
+
+// Get sample rate value that is valid for currently stored results.
+int SaProcessor::getSampleRate() const
+{
+	return m_sampleRate;
+}
+
+
+// Inform the processor whether any display widgets actually need it.
+void SaProcessor::setSpectrumActive(bool active)
+{
+	m_spectrumActive = active;
+}
+
+void SaProcessor::setWaterfallActive(bool active)
+{
+	m_waterfallActive = active;
+}
+
+
+// Reallocate data buffers according to newly set block size.
+void SaProcessor::reallocateBuffers()
+{
+	// The SaProcessor destructor is called before SaControlsDialog destructor.
+	// This causes problems, sice ComboBoxModel emits dataChanged() as it is
+	// destroyed, triggering reallocation in processor that is already gone --
+	// a good recipe for segfault. This check explicitly prevents it.
+	if (m_destroyed) {return;}
```

Comment:
As we've discussed, this is not always safe. I.e. the reallocating thread could have already passed the destroyed check, but not yet have reallocated.

One way to fix this is to remove the destroyed check and instead switch around `SaProcessor` and `SaControls` members in `Analyzer`: The destruction order is reverse to the construction order, which equals the member order. So if `SaControls` get destroyed first, its Combo Boxes are destroyed and cannot fire any signals anymore that the `SaProcessor` must handle, so `SaProcessor` can be safely destroyed afterwards.

A different solution: Before controls or processor are getting destroyed, the destructor of `Analyzer` is being called. There, `Analyzer` could force `SaControls` to disconnect its connections. You my need a mutex, too, to realize that (replacing `m_destroyed`), I guess, but that sounds more clean.

Reply:
I'm all for simple solutions, so I removed m_destroyed and switched the destructor order. I tested it a few times and it seems to work fine, so assuming #5050 is not a problem, as PhysSong suggests, this should be "clean enough". :)

Reply:
OK, let's leave this open until they answer :smile:

Reply:
Actually, I don't think we need to wait, unless you want. Meanwhile, I think #5050 is not a bug. But even if it is, it must be fixed, so your code is still right. Feel free to hit resolve.

Reply:
Well, so far I haven't seen any issues resulting from the change. And if there are any, other plugins would be affected as well, so it should be treated as a separate issue, as you suggest. So resolving.